File: src/face_tracker.cpp
/*
	Last Change : 2024.09.03
*/

#include <opencv2/core/base.hpp>

#include "../include/face_tracker.h"

int FaceTracker::kf_count = 0;

FaceTracker::FaceTracker(): \
            age(0) , time_since_update(0) , num_hits(0) , continual_hits(0)
{
    initialize(cv::Rect_<float>());
    id = kf_count;
    std::cout << "id = kf_count : " << id << std::endl;

}

FaceTracker::FaceTracker(cv::Rect_<float> initRect) : \
            age(0) , time_since_update(0) , num_hits(0) , continual_hits(0)
{
    id = kf_count++;
    initialize(initRect);
}

FaceTracker::~FaceTracker()
{
    state_history.clear();
}

void FaceTracker::initialize(const cv::Rect_<float> stateMat) 
{
    int stateNum = 7;
    int measureNum = 4;
    // Initialize Kalman Filter parameters
    kf.init(stateNum, measureNum , 0); // State dimension, measurement dimension, control dimension
    kf.transitionMatrix = (cv::Mat_<float>(stateNum, stateNum) << 1, 0, 0, 0, 1, 0, 0, \
						   0, 1, 0, 0, 0, 1, 0,
						   0, 0, 1, 0, 0, 0, 1,
						   0, 0, 0, 1, 0, 0, 0,
						   0, 0, 0, 0, 1, 0, 0,
						   0, 0, 0, 0, 0, 1, 0,
						   0, 0, 0, 0, 0, 0, 1);

    setIdentity(kf.measurementMatrix);
    setIdentity(kf.processNoiseCov, cv::Scalar::all(1e-2));
    setIdentity(kf.measurementNoiseCov, cv::Scalar::all(1e-1));
    setIdentity(kf.errorCovPost, cv::Scalar::all(1));

    measurement = cv::Mat::zeros(measureNum, 1, CV_32F);

    // initialize state vector with bounding box in [cx,cy,s,r] style
	kf.statePost.at<float>(0, 0) = stateMat.x + stateMat.width / 2;
	kf.statePost.at<float>(1, 0) = stateMat.y + stateMat.height / 2;
	kf.statePost.at<float>(2, 0) = stateMat.area();
	kf.statePost.at<float>(3, 0) = stateMat.width / stateMat.height;
}

cv::Rect_<float> FaceTracker::xysr2rect(float center_x , float center_y , float s , float r)
{
    float w = sqrt(s * r);
	float h = s / w;
	float x = (center_x - w / 2);
	float y = (center_y - h / 2);

	if (x < 0 && center_x > 0)
		x = 0;
	if (y < 0 && center_y > 0)
		y = 0;

	return cv::Rect_<float>(x, y, w, h);
}

cv::Rect_<float> FaceTracker::predict() 
{   
    cv::Mat prediction = kf.predict();
    age += 1;

    if (time_since_update > 0)
    {
        continual_hits = 0;
    }
    time_since_update += 1;

    cv::Rect_<float> predict_box = xysr2rect( \
        prediction.at<float>(0, 0), prediction.at<float>(1, 0), prediction.at<float>(2, 0), prediction.at<float>(3, 0));

    state_history.push_back(predict_box);

    return state_history.back();
}

void FaceTracker::update(const cv::Rect_<float> new_bbox) 
{
    this->lastState = new_bbox;

    time_since_update = 0;
    state_history.clear();
    num_hits += 1;
    continual_hits += 1;

    // measurement
	measurement.at<float>(0, 0) = new_bbox.x + new_bbox.width / 2;
	measurement.at<float>(1, 0) = new_bbox.y + new_bbox.height / 2;
	measurement.at<float>(2, 0) = new_bbox.area();
	measurement.at<float>(3, 0) = new_bbox.width / new_bbox.height;

    kf.correct(measurement);
}

cv::Rect_<float> FaceTracker::get_state()
{
    cv::Mat state = kf.statePost;
    return xysr2rect(state.at<float>(0,0) , state.at<float>(1,0) , state.at<float>(2,0) , state.at<float>(3,0));
}

File: src/face_tracker.py
import numpy as np
import cv2


class FaceTracker:
    kf_count = 0

    def __init__(self, stateMat=None):
        self.kf = cv2.KalmanFilter(4, 2)  # 状态向量维度为4，测量向量维度为2
        self.measurement = np.zeros((2, 1), np.float32)  # 用以更新的观测矩阵
        self.state_history = []  # 存储历史状态信息

        self.id = FaceTracker.kf_count
        FaceTracker.kf_count += 1
        self.age = 0  # 跟踪器创建后的帧数

        self.currentPosition = None  # 保存当前的跟踪位置
        self.lastState = None  # 存储最后一次更新的矩形状态
        self.time_since_update = 0  # 自上次状态更新以来经过的时间
        self.num_hits = 0  # 跟踪器命中次数，成功匹配到目标的次数
        self.continual_hits = 0  # 连续成功匹配到目标的次数

        if stateMat is not None:
            self.initialize(stateMat)

    def initialize(self, stateMat):
        # 初始化卡尔曼滤波器状态
        self.kf.statePre = np.array([[stateMat[0]], [stateMat[1]], [0], [0]], np.float32)
        self.kf.transitionMatrix = np.array([[1, 0, 1, 0],
                                             [0, 1, 0, 1],
                                             [0, 0, 1, 0],
                                             [0, 0, 0, 1]], np.float32)
        self.kf.measurementMatrix = np.array([[1, 0, 0, 0],
                                              [0, 1, 0, 0]], np.float32)

    def predict(self):
        self.lastState = self.kf.predict()
        return self.lastState

    def update(self, new_bbox):
        self.measurement[0] = new_bbox[0] + new_bbox[2] / 2  # 计算中心点 x
        self.measurement[1] = new_bbox[1] + new_bbox[3] / 2  # 计算中心点 y
        self.kf.correct(self.measurement)

        self.currentPosition = new_bbox
        self.state_history.append(new_bbox)
        self.age += 1

        # 更新跟踪器命中次数
        self.num_hits += 1
        self.continual_hits += 1
        self.time_since_update = 0

    def xysr2rect(self, center_x, center_y, s, r):
        # 根据中心点、宽度和高度计算矩形
        return (center_x - s / 2, center_y - r / 2, s, r)

    def get_state(self):
        return self.lastState
